src/twisted/conch/ssh/transport.py

```diff
@@ -516,6 +515,99 @@ class SSHTransportBase(protocol.Protocol):
     _keyExchangeState = _KEY_EXCHANGE_NONE
     _blockedByKeyExchange = None
 
+
+    def setSupportedKexes(self, kexes):
+        """
+        Clear out the list of supported key exchanges
+        Then call addSupportedKexes to populate the list.
+
+        @type kexes: L{list}
+        @param kexes: A L{list} of kex name strings
+        """
+        # Clear out the currently supported key exchanges.
+        _kex._kexAlgorithms.clear()
+
+        # Set the new exchanges.
+        self.addSupportedKexes(kexes)
+
+
+    def addSupportedKexes(self, kexes):
+        """
+        Append the list of supported key exchanges.
+
+        @type kexes: L{list}
+        @param kexes: A L{list} of kex name strings
+        """
+        for k in kexes:
+            if k.startswith(b'ecdh') and kexes[k] is None:
+                size = int(k[-3:])
+
+                # Add the new kex to the dictionary.
+                if size <= 256:
+                    _kex._kexAlgorithms[k] = _kex._ECDH256()
+                elif size <= 384:
+                    _kex._kexAlgorithms[k] = _kex._ECDH384()
+                else:
+                    _kex._kexAlgorithms[k] = _kex._ECDH512()
+            elif kexes[k] is None:
+                _kex._kexAlgorithms[k] = _kex._baseKexAlgorithms[k]
+            else:
+                _kex._kexAlgorithms[k] = kexes[k]
+
+        # Reload
+        del self.supportedKeyExchanges[:]
+        for kex in _kex.getSupportedKeyExchanges():
+            self.supportedKeyExchanges.append(kex)
+
+
+    def setSupportedPublicKeys(self, publicKeys):
+        """
+         Clear out the list of supported public keys
+         Then call addSupportePublicKeys to populate the list.
+
+         @type publicKeys: L{dict}
+         @param publicKeys: A L{dict} with the name of the public key as the key
+            and a value that is a L{tuple} consisting of an instance of the key method,
+            a L{str} that is the corresponding sec name.
+        """
+        del self.supportedPublicKeys[:]
+        self.addSupportePublicKeys(publicKeys)
+
+
+    def addSupportePublicKeys(self, publicKeys, sort=True):
+        """
+        Populate the supported public key list.
+
+        @type publicKeys: L{dict}
+        @param publicKeys: A L{dict} with the name of the public key as the key
+            and a value that is a L{tuple} consisting of an instance of the key method,
+            a L{str} that is the corresponding sec name.
+        """
+        for k in publicKeys:
+            if k not in self.supportedPublicKeys:
+                self.supportedPublicKeys.append(k)
+
+            if k.startswith(b'ecdsa'):
+                curve = publicKeys[k]
```

Comment:
It is possible, but _curveTable and _secToNist are just reference tables.  Just having a curve in there doesn't mean it's advertised as being supported during negotiations, and I feel that it could cause more issues making them instance variables.